src/vmq_lvldb_store.erl

```diff
@@ -421,14 +417,34 @@ check_delete_message(Bucket, {ok, Key}, Itr, WriteOpts, {PivotMsgRef, PivotMP, H
             %% -> delete required
             eleveldb:write(Bucket, [{delete, Key}], WriteOpts),
             {{NewPivotMsgRef, NewPivotMP, false}, N + 1}; %% challenge the new message
-        {idx, _, _} ->
+        {idx, _, MsgRef} ->
+            incr_ref(Refs, MsgRef),
             %% ignore idx item atm.
             {Pivot, N}
     end,
-    check_delete_message(Bucket, eleveldb:iterator_move(Itr, next), Itr, WriteOpts, NewPivot, NewN);
-check_delete_message(Bucket, {error, invalid_iterator}, _, WriteOpts, {PivotMsgRef, PivotMP, false}, N) ->
+    check_store(Bucket, Refs, eleveldb:iterator_move(Itr, next), Itr, WriteOpts, NewPivot, NewN);
+check_store(Bucket, _, {error, invalid_iterator}, _, WriteOpts, {PivotMsgRef, PivotMP, false}, N) ->
     Key = sext:encode({msg, PivotMsgRef, {PivotMP, ''}}),
     eleveldb:write(Bucket, [{delete, Key}], WriteOpts),
     N + 1;
-check_delete_message(_Bucket, {error, invalid_iterator}, _, _, {_, _, true}, N) ->
+check_store(_Bucket, _, {error, invalid_iterator}, _, _, {_, _, true}, N) ->
     N.
+
+incr_ref(Refs, MsgRef) ->
+    case ets:insert_new(Refs, {MsgRef, 1}) of
+        true -> 1;
+        false ->
+            ets:update_counter(Refs, MsgRef, 1)
+    end.
+
+decr_ref(Refs, MsgRef) ->
+    try ets:update_counter(Refs, MsgRef, -1) of
+        0 ->
+            ets:delete(Refs, MsgRef),
+            0;
+        V ->
+            V
+    catch
+        _:_ ->
+            not_found
+    end.
```

Comment:
Update_counter gives you a badarg if the counter doesn't exist. I think it makes sense to catch this exception and don't crash the msgstore as a restart typically won't help in such error conditions.